Update System.Drawing to reflect GDI+ changes

qmfrederik opened the pull request:
This is the last part of the work done in dotnet/corefx#40181 which was pending API approval (#30543).

1. Add support `ValueTypePointer` encoder parameters ([doc](https://docs.microsoft.com/en-us/windows/win32/api/gdiplusenums/ne-gdiplusenums-encoderparametervaluetype))
2. Support `ColorSpace`, `ImageItems` and `SaveAsCmyk` encoders ([doc](https://docs.microsoft.com/en-us/windows/win32/gdiplus/-gdiplus-constant-image-encoder-constants))
3. By virtue of 1., fixes a bug where `GetEncoderParameterList` was broken for the JPEG encoder, which now returns (an empty list of) parameters of type `EncoderParameterValueTypePointer`. This was broken in .NET Framework, too (see [e.g. this StackOverflow report](https://stackoverflow.com/questions/3152506/unexpected-bitmap-region-is-already-locked-exception-with-getencoderparameterl))

Fixes #30543

qmfrederik committed:
Update System.Drawing to reflect GDI+ changes

- Add support ValueTypePointer encoder parameters
- Support ColorSpace, ImageItems and SaveAsCmyk encoders

qmfrederik committed:
Skip test on NetFx

safern commented on src/libraries/System.Drawing.Common/tests/ImageTests.cs:
```diff
@@ -157,6 +157,23 @@ public static IEnumerable<object[]> GetEncoderParameterList_ReturnsExpected_Test
                 }
             };
+
+#if !NETFRAMEWORK
```
Nit: for tests and test data, we usually use, if (!PlatformDetection.IsFullFramework)

qmfrederik replied:
Hmm,

```csharp
ImageTests.cs(161,36): error CS0117: (NETCORE_ENGINEERING_TELEMETRY=Build) 'PlatformDetection' does not contain a definition for 'IsFullFramework'
```

What am I missing here?

filipnavara replied:
`IsFullFramework`, obivously :)

It was removed in https://github.com/dotnet/corefx/pull/37972.

qmfrederik replied:
Ugh 😄 .

I'm a bit confused, because dotnet/corefx#37972 and dotnet/runtime#28619 seem to indicate tests would no longer run on full framework, yet they do for System.Drawing.Common.

@safern What's the way forward here?

filipnavara replied:
The test should run on `netfx` only if the library targets `netstandard2.0`. In that case they should test the functionality of the `netstandard2.0` library running on NetFX, not testing NetFX code itself.

filipnavara replied:
The tricky issue is the fact that the NuGet package contains `lib/net461` implementation that forwards all the types to `System.Drawing`. I am not entirely sure how is that produced but it is likely problematic going forward.

safern replied:
> IsFullFramework, obviously :)
> It was removed in dotnet/corefx#37972.

It was not removed, it was renamed to `IsNetFramework`. https://github.com/dotnet/runtime/blob/0b25063ec056b0d6668bf0848d108d8c873f3fce/src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/AssertExtensions.cs#L16

https://github.com/dotnet/runtime/pull/1900

> The test should run on netfx only if the library targets netstandard2.0. In that case they should test the functionality of the netstandard2.0 library running on NetFX, not testing NetFX code itself.

Not really. We want to run tests on `netfx` for libraries that ship a netfx asset. In this case even if it is a facade, we want to test that we're producing the right facade and that tests run correctly.

filipnavara replied:
> Not really. We want to run tests on netfx for libraries that ship a netfx asset. 

Fair enough. I thought these ones are mostly in frozen state now and slowly removed from repository with no forward maintenance plan except from release branches.

> In this case even if it is a facade, we want to test that we're producing the right facade and that tests run correctly.

It's somewhat weird to have new APIs that are not present on certain platform. What will happen if you call them? I'd except something like `PlatformNotSupportedException` but that only works for new types and only if partial facade is generated. In this case it's new members on existing types.

safern replied:
> It's somewhat weird to have new APIs that are not present on certain platform. What will happen if you call them? I'd except something like PlatformNotSupportedException but that only works for new types and only if partial facade is generated. In this case it's new members on existing type.

You will not be able to reference them anymore as for ref we only build live for netcoreapp3.0 and netcoreapp5.0. The netfx assets harvested. So now that you mention it we no longer build netfx assets for System.Drawing.Common anymore as we just get them from the latest package and re-ship  them in the new package. Maybe it is fine removing the netfx tests for this library.

cc: @ericstj @ViktorHofer @danmosemsft  @JeremyKuhne -- I'm leaning towards removing testing netfx for S.Drawing.Common. What do you guys think?

JeremyKuhne replied:
> What do you guys think?

My only concern is that surfacing unintentional deviation in behavior from Framework becomes more difficult without runs there. If we had better test coverage I'd worry less, but I think we're pretty far from adequately covering GDI+ behavior in our current tests.

safern replied:
Yeah that's a fair point. I would suggest leaving the tests on this PR and using the `NetFramework` property in `PlatformDetection`; then we can discuss further if we want to stop running tests for netfx in this project.

qmfrederik replied:
So, unfortunately we went the full circle on this one - just using `PlatformDetection` does not work, because there's a reference to `Encoder.ImageItems` which was added as part of this PR, but is not present in NetFx.

safern replied:
You're right. Ok, well this at least was a good discussion. Then we will need to go back to the compile time condition, sorry about that.

safern reviewed (approved):
Other than a comment looks good to me. 

cc: @JeremyKuhne

stephentoub commented:
@qmfrederik, if you can address @safern's feedback, seems like we can then get this merged. Thanks.

stephentoub commented on src/libraries/System.Drawing.Common/ref/System.Drawing.Common.cs:
```diff
@@ -2105,6 +2105,9 @@ public sealed partial class Encoder
         public static readonly System.Drawing.Imaging.Encoder Version;
+        public static readonly System.Drawing.Imaging.Encoder ColorSpace;
+        public static readonly System.Drawing.Imaging.Encoder ImageItems;
+        public static readonly System.Drawing.Imaging.Encoder SaveAsCmyk;
```
@safern, @carlossanlop, I assume we want to add XML comments to these in order to populate the docs?  For consistency maybe we should do so for all the fields here.

stephentoub commented on src/libraries/System.Drawing.Common/ref/System.Drawing.Common.cs:
```diff
@@ -2152,6 +2155,7 @@ public enum EncoderParameterValueType
         ValueTypeLongRange = 6,
         ValueTypeUndefined = 7,
         ValueTypeRationalRange = 8,
+        ValueTypePointer = 9,
```
Ditto

stephentoub commented on src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Encoder.cs:
```diff
@@ -16,6 +16,9 @@ public sealed class Encoder
         public static readonly Encoder SaveFlag = new Encoder(new Guid(unchecked((int)0x292266fc), unchecked((short)0xac40), unchecked((short)0x47bf), new byte[] { 0x8c, 0xfc, 0xa8, 0x5b, 0x89, 0xa6, 0x55, 0xde }));
+        public static readonly Encoder ColorSpace = new Encoder(new Guid(unchecked((int)0xae7a62a0), unchecked((short)0xee2c), unchecked((short)0x49d8), new byte[] { 0x9d, 0x07, 0x1b, 0xa8, 0xa9, 0x27, 0x59, 0x6e }));
+        public static readonly Encoder ImageItems = new Encoder(new Guid(unchecked((int)0x63875e13), unchecked((short)0x1f1d), unchecked((short)0x45ab), new byte[] { 0x91, 0x95, 0xa2, 0x9b, 0x60, 0x66, 0xa6, 0x50 }));
+        public static readonly Encoder SaveAsCmyk = new Encoder(new Guid(unchecked((int)0xa219bbc9), unchecked((short)0x0a9d), unchecked((short)0x4005), new byte[] { 0xa3, 0xee, 0x3a, 0x42, 0x1b, 0x8b, 0xb0, 0x6c }));
```
@safern, @carlossanlop, I assume we want to add XML comments to these in order to populate the docs? For consistency maybe we should do so for all the fields here.

safern replied:
@stephentoub, good catch. @qmfrederik could you please also include XML docs for these new APIs?

carlossanlop replied:
@qmfrederik the only information that the existing APIs need on top of them as triple slash is their summary, which you can get from their [MS Docs documentationi](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.imaging.encoder?view=netframework-4.8).

qmfrederik committed:
Add XML documentation

qmfrederik commented:
OK, I think I addressed the feedback.

The XML comments keep the same format used in the previous versions of the framework. It doesn't look like GDI+ has docs with an elaborate description of these encoder parameters, so I left it at that.

safern commented:
@carlossanlop could you sign off on the XML comments?

carlossanlop commented on src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Encoder.cs:
```diff
@@ -16,6 +16,21 @@ public sealed class Encoder
         public static readonly Encoder SaveFlag = new Encoder(new Guid(unchecked((int)0x292266fc), unchecked((short)0xac40), unchecked((short)0x47bf), new byte[] { 0x8c, 0xfc, 0xa8, 0x5b, 0x89, 0xa6, 0x55, 0xde }));
+        
+        /// <summary>
+        ///	Gets an Encoder object that is initialized with the globally unique identifier for the color space category.
```
Two requests: remove the "Gets" at the beginning, which we normally only use for properties with a getter. And add a xref link to the Encoder type. Like this:

```suggestion
        ///	An <xref:System.Drawing.Imaging.Encoder> object that is initialized with the globally unique identifier for the color space category.
```

carlossanlop commented on src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Encoder.cs:
```diff
@@ -16,6 +16,21 @@ public sealed class Encoder
+        public static readonly Encoder ColorSpace = new Encoder(new Guid(unchecked((int)0xae7a62a0), unchecked((short)0xee2c), unchecked((short)0x49d8), new byte[] { 0x9d, 0x07, 0x1b, 0xa8, 0xa9, 0x27, 0x59, 0x6e }));
+        
+        /// <summary>
+        ///	Gets an Encoder object that is initialized with the globally unique identifier for the image items category.
```
Same:
```suggestion
        ///	An <see cref="T:System.Drawing.Imaging.Encoder" /> object that is initialized with the globally unique identifier for the image items category.
```

carlossanlop commented on src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Encoder.cs:
```diff
@@ -16,6 +16,21 @@ public sealed class Encoder
+        public static readonly Encoder ImageItems = new Encoder(new Guid(unchecked((int)0x63875e13), unchecked((short)0x1f1d), unchecked((short)0x45ab), new byte[] { 0x91, 0x95, 0xa2, 0x9b, 0x60, 0x66, 0xa6, 0x50 }));
+        
+        /// <summary>
+        ///	Gets an Encoder object that is initialized with the globally unique identifier for the save as CMYK category.
```
Same:
```suggestion
        ///	An <see cref="T:System.Drawing.Imaging.Encoder" /> object that is initialized with the globally unique identifier for the save as CMYK category.
```

carlossanlop reviewed (approved):
Looks good for documentation, @qmfrederik. But just for consistency with the existing documentation in MS Docs, please consider applying the suggestions I offered.

safern commented on src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Encoder.cs:
```diff
@@ -16,6 +16,21 @@ public sealed class Encoder
+        public static readonly Encoder ImageItems = new Encoder(new Guid(unchecked((int)0x63875e13), unchecked((short)0x1f1d), unchecked((short)0x45ab), new byte[] { 0x91, 0x95, 0xa2, 0x9b, 0x60, 0x66, 0xa6, 0x50 }));
+        
+        /// <summary>
+        /// An <see cref="Encoder" /> object that is initialized with the globally unique identifier for the save as CMYK category.
```
I think we try to use the full name even if it is not needed, but not sure. @carlossanlop ?

carlossanlop replied:
Do you mean using `<see cref="Encoder" />`? No, it's not needed in triple slash comments. If the API was not recognized within this context, VS would give you an error. But in this case, we recognize the `Encoder` type.
When the project gets built, the triple slash comments get converted to the xml files, which will automatically convert types to their full name.

qmfrederik replied:
Using the `T:` prefix results in build errors: `error CA1200: Avoid using cref tags with a prefix`

qmfrederik commented on src/libraries/System.Drawing.Common/tests/ImageTests.cs:
```diff
@@ -157,6 +157,24 @@ public static IEnumerable<object[]> GetEncoderParameterList_ReturnsExpected_Test
+                        Encoder.Quality.Guid,
+                        Encoder.LuminanceTable.Guid,
+                        Encoder.ChrominanceTable.Guid,
+                        Encoder.ImageItems.Guid
```
So, the `ImageItems` property was added in this PR, but is not present in the `NetFramework` builds since they type forward to System.Drawing, which does not have this property.

How should we tackle this?

qmfrederik commented:
/azp help

qmfrederik commented:
/azp run "runtime (Installer Build and Test Linux_x64 Debug)"

qmfrederik commented:
The build failures seem unrelated but I don't have sufficient permissions to trigger a rerun.

safern commented:
I re-ran them. You can always go to the checks tab and click on re-run no?

qmfrederik commented:
Nope, I don't have those permissions. #718 has some more information.

Looking forward to a green CI and getting this merged!
